### PR TITLE
observability

### PR DESCRIPTION
**Pull request by PietroPasotti**
This PR adds loki-logging and grafana-dashboarding integrations to the traefik charm.

## Testing Instructions
deploy cos-lite, then refresh the traefik charm with one built from these sources. Visit the grafana frontend to see the traefik dashboard with the logs.

## Release Notes
- Added loki-logging and grafana-dashboarding integrations.

### CODE REVIEW

**Comment by PietroPasotti** [comment]
To be determined:
- we have three request processing duration traces: a histogram per service, per router, or per entrypoint (it's three separate histograms). Do we want to show them separately in the dashboard, combine them somehow, or show just one of them (which one?)  For now, I'm only showing the per-entrypoint histogram.
- Names for the two custom panels I added -- 'raw logs' and 'long tail'
- Names for the legend of the 'long tail' graph traces.

**Comment by lucabello** [comment]
Grafana errors are preventing me to have a look at the dashboards; in any case, I have rebased on main !

File: metadata.yaml [hunk 1]
```diff
@@ -53,4 +53,15 @@ provides:
...
+  grafana-dashboard:
+    description: |
+      Provides endpoint to instrument a grafana dashboard upon relating to a Grafana charm.
+    interface: grafana_dashboard
+
+requires:
+  logging:
+    interface: loki_push_api
+    description: |
+      Endpoint to push logs to Loki.
```
Comment (sed-i) [inline metadata.yaml:66]:
```suggestion
      Receives Loki's push api endpoint address to push logs to, and forwards charm's built-in alert rules to Loki.
```

File: src/charm.py [hunk 2]
```diff
@@ -93,6 +96,18 @@ def __init__(self, *args):
             ports=[(f"{self.app.name}", self._port)],
         )
 
+        # observability endpoints
```
Comment (sed-i) [inline src/charm.py:99]: I think the correct term here is integrations (formerly: relations)
```suggestion
        # Observability integrations
```

**Review by sed-i — APPROVED** [review]

**Review by Abuelodelanada — COMMENTED** [review]
If I relate `traefik` with `grafana-k8s` `edge` Rev: `58`, I get the following error in Grafana:

```
unit-grafana-0: 10:33:24.379 ERROR unit.grafana/0.juju-log grafana-dashboard:41: Uncaught exception while in charm code:
Traceback (most recent call last):
  File "./src/charm.py", line 1167, in <module>
    main(GrafanaCharm, use_juju_for_storage=True)
  File "/var/lib/juju/agents/unit-grafana-0/charm/venv/ops/main.py", line 441, in main
    _emit_charm_event(charm, dispatcher.event_name)
  File "/var/lib/juju/agents/unit-grafana-0/charm/venv/ops/main.py", line 145, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-grafana-0/charm/venv/ops/framework.py", line 355, in emit
    framework._emit(event)  # noqa
  File "/var/lib/juju/agents/unit-grafana-0/charm/venv/ops/framework.py", line 856, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-grafana-0/charm/venv/ops/framework.py", line 931, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-grafana-0/charm/lib/charms/grafana_k8s/v0/grafana_dashboard.py", line 1337, in _on_grafana_dashboard_relation_changed
    changes = self._render_dashboards_and_signal_changed(event.relation)
  File "/var/lib/juju/agents/unit-grafana-0/charm/lib/charms/grafana_k8s/v0/grafana_dashboard.py", line 1434, in _render_dashboards_and_signal_changed
    content = _convert_dashboard_fields(content, inject_dropdowns)
  File "/var/lib/juju/agents/unit-grafana-0/charm/lib/charms/grafana_k8s/v0/grafana_dashboard.py", line 599, in _convert_dashboard_fields
    dict_content = _replace_template_fields(dict_content, datasources, existing_templates)
  File "/var/lib/juju/agents/unit-grafana-0/charm/lib/charms/grafana_k8s/v0/grafana_dashboard.py", line 618, in _replace_template_fields
    dict_content["panels"] = _template_panels(
  File "/var/lib/juju/agents/unit-grafana-0/charm/lib/charms/grafana_k8s/v0/grafana_dashboard.py", line 693, in _template_panels
    replacement = replacements.get(datasources[ds], "")
KeyError: 'grafana'
unit-traefik-0: 10:33:24.708 INFO juju.worker.uniter.operation ran "grafana-dashboard-relation-changed" hook (via hook dispatching script: dispatch)
```

**Comment by PietroPasotti** [comment]
> If I relate `traefik` with `grafana-k8s` `edge` Rev: `58`, I get the following error in Grafana:
> 
> ```
> unit-grafana-0: 10:33:24.379 ERROR unit.grafana/0.juju-log grafana-dashboard:41: Uncaught exception while in charm code:
> Traceback (most recent call last):
>   File "./src/charm.py", line 1167, in <module>
>     main(GrafanaCharm, use_juju_for_storage=True)
>   File "/var/lib/juju/agents/unit-grafana-0/charm/venv/ops/main.py", line 441, in main
>     _emit_charm_event(charm, dispatcher.event_name)
>   File "/var/lib/juju/agents/unit-grafana-0/charm/venv/ops/main.py", line 145, in _emit_charm_event
>     event_to_emit.emit(*args, **kwargs)
>   File "/var/lib/juju/agents/unit-grafana-0/charm/venv/ops/framework.py", line 355, in emit
>     framework._emit(event)  # noqa
>   File "/var/lib/juju/agents/unit-grafana-0/charm/venv/ops/framework.py", line 856, in _emit
>     self._reemit(event_path)
>   File "/var/lib/juju/agents/unit-grafana-0/charm/venv/ops/framework.py", line 931, in _reemit
>     custom_handler(event)
>   File "/var/lib/juju/agents/unit-grafana-0/charm/lib/charms/grafana_k8s/v0/grafana_dashboard.py", line 1337, in _on_grafana_dashboard_relation_changed
>     changes = self._render_dashboards_and_signal_changed(event.relation)
>   File "/var/lib/juju/agents/unit-grafana-0/charm/lib/charms/grafana_k8s/v0/grafana_dashboard.py", line 1434, in _render_dashboards_and_signal_changed
>     content = _convert_dashboard_fields(content, inject_dropdowns)
>   File "/var/lib/juju/agents/unit-grafana-0/charm/lib/charms/grafana_k8s/v0/grafana_dashboard.py", line 599, in _convert_dashboard_fields
>     dict_content = _replace_template_fields(dict_content, datasources, existing_templates)
>   File "/var/lib/juju/agents/unit-grafana-0/charm/lib/charms/grafana_k8s/v0/grafana_dashboard.py", line 618, in _replace_template_fields
>     dict_content["panels"] = _template_panels(
>   File "/var/lib/juju/agents/unit-grafana-0/charm/lib/charms/grafana_k8s/v0/grafana_dashboard.py", line 693, in _template_panels
>     replacement = replacements.get(datasources[ds], "")
> KeyError: 'grafana'
> unit-traefik-0: 10:33:24.708 INFO juju.worker.uniter.operation ran "grafana-dashboard-relation-changed" hook (via hook dispatching script: dispatch)
> ```

---> https://github.com/canonical/grafana-k8s-operator/issues/173

File: src/loki_alert_rules/error_level_logs_found.rule [hunk 3]
```diff
@@ -0,0 +1,7 @@
+alert: ErrorLevelLogsFound
+expr: rate({%%juju_topology%%} |= "level=error"[1m]) > 0
```
Comment (Abuelodelanada) [inline src/loki_alert_rules/error_level_logs_found.rule:2]: Unlike Prometheus alert rules, in Loki based alert rules `%%juju_topology%%` is needed to [be replaced](https://github.com/canonical/loki-k8s-operator/blob/main/lib/charms/loki_k8s/v0/loki_push_api.py#L768):

```python
alert_rule["expr"] = self.tool.inject_label_matchers(
                            re.sub(r"%%juju_topology%%", r'job=".+"', alert_rule["expr"]),
                            self.topology.label_matcher_dict,
                        )
``` 

At this point this `exp` will be:

```
rate({job=".+"} |= "level=error"[1m]) > 0
```

After that juju_topology is added using `cos-tool`, so the `expr` will be:

```
(rate({job=".+", juju_application="traefik", juju_charm="traefik-k8s", juju_model="cos",
      juju_model_uuid="381236d6-8c63-4415-80ad-bfdccc1ddb4c"} |= "level=error"[1m])
      > 0)
```

**Review by Abuelodelanada — APPROVED** [review]

**Comment by PietroPasotti** [comment]
closing this PR for now because the dashboard is broken and I'm busy fixing it. besides, it's getting messy and the commits need some cleaning up

**Pull request closed without merging**